cpp/src/slice2cs/Gen.cpp

```diff
@@ -2025,49 +2011,95 @@ Slice::Gen::TypesVisitor::visitClassDefEnd(const ClassDefPtr& p)
     {
         _out << sp << nl << "#region Constructors";
 
-        _out << sp;
-        emitGeneratedCodeAttribute();
-        _out << nl << "public " << name << spar << epar;
-        if (base)
-        {
-            _out << " : base()";
-        }
-        _out << sb;
-        writeDataMemberInitializers(dataMembers, DotNet::ICloneable);
-        _out << nl << "ice_initialize();";
-        _out << eb;
-
+        // Primary constructor.
         _out << sp;
         emitGeneratedCodeAttribute();
         _out << nl << "public " << name << spar;
+
         vector<string> paramDecl;
-        for (DataMemberList::const_iterator d = allDataMembers.begin(); d != allDataMembers.end(); ++d)
+        vector<string> secondaryCtorParams;
+        vector<string> secondaryCtorMemberNames;
+        vector<string> secondaryCtorBaseParamNames;
+        for (const auto& q : allDataMembers)
         {
-            string memberName = fixId((*d)->name(), DotNet::ICloneable);
-            string memberType = typeToString((*d)->type(), ns, (*d)->optional());
+            string memberName = fixId(q->name(), DotNet::ICloneable);
+            string memberType = typeToString(q->type(), ns, q->optional());
             paramDecl.push_back(memberType + " " + memberName);
+
+            StructPtr st = dynamic_pointer_cast<Struct>(q->type());
+            if (dynamic_pointer_cast<Sequence>(q->type()) || dynamic_pointer_cast<Dictionary>(q->type()) ||
```

Comment:
Why don't we check for optional here?

Reply:
You're right, we should exclude optional data members here. Fixed.